File: backend/physics/forces.py
from __future__ import annotations
from dataclasses import dataclass
import torch
from .particle import ParticleSystem


@dataclass
class SpringSystem:
    """Force-based Hooke's law springs with damping.

    Spring force formula (Paul Bourke):
        F = [ks * (|xb - xa| - r) + kd * dot(vb - va, n)] * n
    where n = unit vector from a to b, ks = spring constant, kd = damping,
    r = rest length. Force on a is +F, force on b is -F (Newton's 3rd law).

    Reference: https://paulbourke.net/miscellaneous/particle/
    """

    indices_a: torch.Tensor     # (S,) int64
    indices_b: torch.Tensor     # (S,) int64
    rest_lengths: torch.Tensor  # (S,) float32
    stiffnesses: torch.Tensor   # (S,) float32
    dampings: torch.Tensor      # (S,) float32

    @staticmethod
    def create() -> SpringSystem:
        return SpringSystem(
            indices_a=torch.zeros(0, dtype=torch.long),
            indices_b=torch.zeros(0, dtype=torch.long),
            rest_lengths=torch.zeros(0),
            stiffnesses=torch.zeros(0),
            dampings=torch.zeros(0),
        )

    @property
    def count(self) -> int:
        return self.indices_a.shape[0]

    def add(self, a: int, b: int, rest_length: float,
            stiffness: float = 0.1, damping: float = 0.01) -> None:
        self.indices_a = torch.cat([self.indices_a, torch.tensor([a], dtype=torch.long)])
        self.indices_b = torch.cat([self.indices_b, torch.tensor([b], dtype=torch.long)])
        self.rest_lengths = torch.cat([self.rest_lengths, torch.tensor([rest_length])])
        self.stiffnesses = torch.cat([self.stiffnesses, torch.tensor([stiffness])])
        self.dampings = torch.cat([self.dampings, torch.tensor([damping])])


def apply_spring_forces(ps: ParticleSystem, springs: SpringSystem) -> None:
    """Apply Hooke's law spring forces with velocity-based damping along the spring axis.

    For each spring connecting particles a and b:
      d = pos_b - pos_a           (displacement vector)
      dist = |d|
      n = d / dist                (unit direction a -> b)
      deform = dist - rest_length (extension > 0, compression < 0)
      F_spring = ks * deform      (Hooke's law, scalar)
      F_damp = kd * dot(vel_b - vel_a, n)  (damping along spring axis)
      F = (F_spring + F_damp) * n (total force vector)
      particle_a gets +F, particle_b gets -F

    References:
      - https://paulbourke.net/miscellaneous/particle/
      - https://www.gorillasun.de/blog/spring-physics-and-connecting-particles-with-springs/
    """
    if springs.count == 0:
        return

    pos_a = ps.positions[springs.indices_a]  # (S, dim)
    pos_b = ps.positions[springs.indices_b]  # (S, dim)
    vel_a = ps.velocities[springs.indices_a]  # (S, dim)
    vel_b = ps.velocities[springs.indices_b]  # (S, dim)

    d = pos_b - pos_a                                       # (S, dim)
    dist = d.norm(dim=-1, keepdim=True).clamp(min=1e-6)     # (S, 1)
    n = d / dist                                             # (S, 1) unit direction

    # Hooke's law: F_spring = ks * (dist - rest_length)
    deform = dist - springs.rest_lengths.unsqueeze(-1)       # (S, 1)
    f_spring = springs.stiffnesses.unsqueeze(-1) * deform    # (S, 1)

    # Damping along spring axis: F_damp = kd * dot(v_rel, n)
    v_rel = vel_b - vel_a                                    # (S, dim)
    v_along = (v_rel * n).sum(dim=-1, keepdim=True)          # (S, 1) scalar projection
    f_damp = springs.dampings.unsqueeze(-1) * v_along        # (S, 1)

    # Total force vector along the spring direction
    force = (f_spring + f_damp) * n                          # (S, dim)

    # Apply: a gets +F (pulled toward b), b gets -F (Newton's 3rd law)
    # Zero out forces for pinned particles via inv_mass
    inv_mass_a = ps.inv_masses[springs.indices_a].unsqueeze(-1)
    inv_mass_b = ps.inv_masses[springs.indices_b].unsqueeze(-1)

    acc_a = force * inv_mass_a
    acc_b = force * inv_mass_b

    ps.accelerations.index_add_(0, springs.indices_a, acc_a)
    ps.accelerations.index_add_(0, springs.indices_b, -acc_b)


@dataclass
class AttractionSystem:
    """Inverse-square attraction/repulsion forces between particle pairs.
    F = strength / |r|^2 along the line joining the particles.
    Reference: https://paulbourke.net/miscellaneous/particle/"""

    indices_a: torch.Tensor   # (A,) int64
    indices_b: torch.Tensor   # (A,) int64
    strengths: torch.Tensor   # (A,) float32
    min_dists: torch.Tensor   # (A,) float32
    max_dists: torch.Tensor   # (A,) float32

    @staticmethod
    def create() -> AttractionSystem:
        return AttractionSystem(
            indices_a=torch.zeros(0, dtype=torch.long),
            indices_b=torch.zeros(0, dtype=torch.long),
            strengths=torch.zeros(0),
            min_dists=torch.zeros(0),
            max_dists=torch.zeros(0),
        )

    @property
    def count(self) -> int:
        return self.indices_a.shape[0]

    def add(self, a: int, b: int, strength: float = 100.0,
            min_dist: float = 20.0, max_dist: float = 500.0) -> None:
        self.indices_a = torch.cat([self.indices_a, torch.tensor([a], dtype=torch.long)])
        self.indices_b = torch.cat([self.indices_b, torch.tensor([b], dtype=torch.long)])
        self.strengths = torch.cat([self.strengths, torch.tensor([strength])])
        self.min_dists = torch.cat([self.min_dists, torch.tensor([min_dist])])
        self.max_dists = torch.cat([self.max_dists, torch.tensor([max_dist])])


def apply_attractions(ps: ParticleSystem, attractions: AttractionSystem) -> None:
    """Apply inverse-square attraction forces between particle pairs."""
    if attractions.count == 0:
        return

    pos_a = ps.positions[attractions.indices_a]
    pos_b = ps.positions[attractions.indices_b]
    delta = pos_b - pos_a
    dist_sq = (delta * delta).sum(dim=-1)  # (A,)
    dist = dist_sq.sqrt()

    # Mask out pairs outside [min_dist, max_dist] or too close
    valid = (dist > 1e-3) & (dist >= attractions.min_dists) & (dist <= attractions.max_dists)

    # Normalize delta
    direction = delta / dist.unsqueeze(-1).clamp(min=1e-6)
    strength = attractions.strengths / dist_sq.clamp(min=1e-6)  # (A,)
    force = direction * strength.unsqueeze(-1)  # (A, dim)

    # Zero out invalid forces
    force = torch.where(valid.unsqueeze(-1), force, torch.zeros_like(force))

    # Apply: a attracted toward b, b repelled from a
    inv_mass_a = ps.inv_masses[attractions.indices_a].unsqueeze(-1)
    inv_mass_b = ps.inv_masses[attractions.indices_b].unsqueeze(-1)

    ps.accelerations.index_add_(0, attractions.indices_a, force * inv_mass_a)
    ps.accelerations.index_add_(0, attractions.indices_b, -force * inv_mass_b)


def apply_viscous_drag(ps: ParticleSystem, coefficient: float) -> None:
    """Viscous drag: F = -kd * v. Applied as acceleration = -kd * v * inv_mass.
    Reference: https://paulbourke.net/miscellaneous/particle/"""
    if coefficient <= 0:
        return
    drag_force = -coefficient * ps.velocities  # (N, dim)
    ps.accelerations += drag_force * ps.inv_masses.unsqueeze(-1)


def apply_gravity(ps: ParticleSystem, gravity: torch.Tensor) -> None:
    """Apply uniform gravitational acceleration. gravity is (dim,) tensor.
    F = m * g => a = g (for non-pinned, where inv_mass > 0).
    Reference: https://paulbourke.net/miscellaneous/particle/"""
    # a = F/m = (m*g)/m = g, but we need inv_mass to zero out pinned particles
    # So: a += g * (inv_mass > 0)
    active = (ps.inv_masses > 0).unsqueeze(-1).float()
    ps.accelerations += gravity.unsqueeze(0) * active


def apply_wind(ps: ParticleSystem, wind_strength: float) -> None:
    """Random horizontal wind noise as a force."""
    if wind_strength <= 0:
        return
    wind = torch.zeros(ps.count, ps.dim)
    wind[:, 0] = wind_strength * (torch.rand(ps.count) - 0.5)
    ps.accelerations += wind * ps.inv_masses.unsqueeze(-1)

File: frontend/src/physics/forces.ts
import type { ParticleSystem } from "./ParticleSystem";

export class SpringSystem {
  indicesA: Uint16Array;
  indicesB: Uint16Array;
  restLengths: Float32Array;
  stiffnesses: Float32Array;
  dampings: Float32Array;
  count: number;

  constructor() {
    this.indicesA = new Uint16Array(0);
    this.indicesB = new Uint16Array(0);
    this.restLengths = new Float32Array(0);
    this.stiffnesses = new Float32Array(0);
    this.dampings = new Float32Array(0);
    this.count = 0;
  }

  add(a: number, b: number, restLength: number, stiffness: number, damping: number = 0): void {
    const i = this.count;
    this.count++;

    const grow = <T extends Uint16Array | Float32Array>(
      arr: T,
      val: number,
    ): T => {
      const Constructor = arr.constructor as { new (len: number): T };
      const next = new Constructor(this.count);
      next.set(arr);
      next[i] = val;
      return next;
    };

    this.indicesA = grow(this.indicesA, a);
    this.indicesB = grow(this.indicesB, b);
    this.restLengths = grow(this.restLengths, restLength);
    this.stiffnesses = grow(this.stiffnesses, stiffness);
    this.dampings = grow(this.dampings, damping);
  }
}

/**
 * Position-based spring correction (Verlet mode).
 * Directly moves particles toward rest length — no force accumulation.
 * Matches traerphysics.js Spring.apply():
 *   diff = (dist - rest) / dist
 *   correction = k * diff * 0.5 * delta
 *   a.position += correction; b.position -= correction
 */
export function applySpringsPositionBased(ps: ParticleSystem, springs: SpringSystem): void {
  if (springs.count === 0) return;
  const dim = ps.dim;
  const pos = ps.positions;

  for (let s = 0; s < springs.count; s++) {
    const a = springs.indicesA[s];
    const b = springs.indicesB[s];
    const k = springs.stiffnesses[s];
    const rest = springs.restLengths[s];

    let distSq = 0;
    for (let d = 0; d < dim; d++) {
      const dd = pos[b * dim + d] - pos[a * dim + d];
      distSq += dd * dd;
    }
    const dist = Math.sqrt(distSq) || 1;
    const diff = (dist - rest) / dist;

    const pinnedA = ps.pinned[a];
    const pinnedB = ps.pinned[b];

    for (let d = 0; d < dim; d++) {
      const delta = pos[b * dim + d] - pos[a * dim + d];
      const correction = k * diff * 0.5 * delta;
      if (!pinnedA) pos[a * dim + d] += correction;
      if (!pinnedB) pos[b * dim + d] -= correction;
    }
  }
}

/**
 * Force-based Hooke's law springs with velocity damping (Euler mode).
 * Used by mesh3d where explicit velocity is needed for collisions.
 * F = [ks * (dist - rest) + kd * dot(vrel, n)] * n
 */
export function applySpringForces(ps: ParticleSystem, springs: SpringSystem): void {
  if (springs.count === 0) return;
  const dim = ps.dim;
  const pos = ps.positions;
  const vel = ps.velocities;
  const acc = ps.accelerations;
  const invM = ps.invMasses;

  for (let s = 0; s < springs.count; s++) {
    const a = springs.indicesA[s];
    const b = springs.indicesB[s];
    const ks = springs.stiffnesses[s];
    const kd = springs.dampings[s];
    const rest = springs.restLengths[s];

    let distSq = 0;
    for (let d = 0; d < dim; d++) {
      const dd = pos[b * dim + d] - pos[a * dim + d];
      distSq += dd * dd;
    }
    const dist = Math.sqrt(distSq);
    if (dist < 1e-6) continue;
    const invDist = 1.0 / dist;

    const deform = dist - rest;
    const fSpring = ks * deform;

    let vAlong = 0;
    for (let d = 0; d < dim; d++) {
      const n = (pos[b * dim + d] - pos[a * dim + d]) * invDist;
      vAlong += (vel[b * dim + d] - vel[a * dim + d]) * n;
    }
    const fTotal = fSpring + kd * vAlong;

    for (let d = 0; d < dim; d++) {
      const n = (pos[b * dim + d] - pos[a * dim + d]) * invDist;
      const f = fTotal * n;
      acc[a * dim + d] += f * invM[a];
      acc[b * dim + d] -= f * invM[b];
    }
  }
}

export class AttractionSystem {
  indicesA: Uint16Array;
  indicesB: Uint16Array;
  strengths: Float32Array;
  minDists: Float32Array;
  maxDists: Float32Array;
  count: number;

  constructor() {
    this.indicesA = new Uint16Array(0);
    this.indicesB = new Uint16Array(0);
    this.strengths = new Float32Array(0);
    this.minDists = new Float32Array(0);
    this.maxDists = new Float32Array(0);
    this.count = 0;
  }

  add(a: number, b: number, strength: number, minDist: number, maxDist: number): void {
    const i = this.count;
    this.count++;

    const grow = <T extends Uint16Array | Float32Array>(
      arr: T,
      val: number,
    ): T => {
      const Constructor = arr.constructor as { new (len: number): T };
      const next = new Constructor(this.count);
      next.set(arr);
      next[i] = val;
      return next;
    };

    this.indicesA = grow(this.indicesA, a);
    this.indicesB = grow(this.indicesB, b);
    this.strengths = grow(this.strengths, strength);
    this.minDists = grow(this.minDists, minDist);
    this.maxDists = grow(this.maxDists, maxDist);
  }
}

export function applyAttractions(ps: ParticleSystem, attractions: AttractionSystem): void {
  if (attractions.count === 0) return;
  const dim = ps.dim;
  const pos = ps.positions;
  const acc = ps.accelerations;
  const invM = ps.invMasses;

  for (let s = 0; s < attractions.count; s++) {
    const a = attractions.indicesA[s];
    const b = attractions.indicesB[s];

    let distSq = 0;
    for (let d = 0; d < dim; d++) {
      const dd = pos[b * dim + d] - pos[a * dim + d];
      distSq += dd * dd;
    }
    const dist = Math.sqrt(distSq);
    if (dist < 1e-3) continue;
    if (dist < attractions.minDists[s] || dist > attractions.maxDists[s]) continue;

    const strength = attractions.strengths[s] / Math.max(distSq, 1e-6);

    for (let d = 0; d < dim; d++) {
      const direction = (pos[b * dim + d] - pos[a * dim + d]) / dist;
      const f = direction * strength;
      acc[a * dim + d] += f * invM[a];
      acc[b * dim + d] -= f * invM[b];
    }
  }
}

/**
 * Viscous drag for Euler mode. Uses explicit velocity.
 * F = -c * v, with clamping to prevent overdamping instability.
 * The effective coefficient is limited so drag removes at most 90% of velocity per frame.
 */
export function applyViscousDrag(ps: ParticleSystem, coefficient: number, dt: number): void {
  if (coefficient <= 0) return;
  const dim = ps.dim;
  for (let i = 0; i < ps.count; i++) {
    const im = ps.invMasses[i];
    if (im <= 0) continue;
    // Clamp: c * im * dt < 0.9 → c < 0.9 / (im * dt)
    const effectiveC = Math.min(coefficient, 0.9 / (im * dt));
    for (let d = 0; d < dim; d++) {
      ps.accelerations[i * dim + d] += -effectiveC * ps.velocities[i * dim + d] * im;
    }
  }
}

/**
 * Viscous drag for Verlet mode. Derives velocity from pos - prev.
 * Matches traerphysics.js Drag.apply():
 *   vel = pos - prev; F = -c * vel; applyForce(F)
 */
export function applyVerletDrag(ps: ParticleSystem, coefficient: number): void {
  if (coefficient <= 0) return;
  const dim = ps.dim;
  const pos = ps.positions;
  const prev = ps.prevPositions;
  const acc = ps.accelerations;
  for (let i = 0; i < ps.count; i++) {
    const im = ps.invMasses[i];
    if (im <= 0) continue;
    for (let d = 0; d < dim; d++) {
      const vel = pos[i * dim + d] - prev[i * dim + d];
      acc[i * dim + d] += -coefficient * vel * im;
    }
  }
}

export function applyGravity(ps: ParticleSystem, gravity: Float32Array): void {
  const dim = ps.dim;
  for (let i = 0; i < ps.count; i++) {
    if (ps.invMasses[i] <= 0) continue;
    for (let d = 0; d < dim; d++) {
      ps.accelerations[i * dim + d] += gravity[d];
    }
  }
}

export function applyWind(ps: ParticleSystem, windStrength: number): void {
  if (windStrength <= 0) return;
  const dim = ps.dim;
  for (let i = 0; i < ps.count; i++) {
    const im = ps.invMasses[i];
    if (im <= 0) continue;
    ps.accelerations[i * dim] += windStrength * (Math.random() - 0.5) * im;
  }
}
